staging/src/k8s.io/apimachinery/pkg/api/validate/net.go

```diff
@@ -17,20 +17,29 @@ limitations under the License.
 package validate
 
 import (
+	"net"
+
 	"k8s.io/apimachinery/pkg/api/operation"
 	"k8s.io/apimachinery/pkg/util/validation/field"
 	netutils "k8s.io/utils/net"
 )
 
-// IP verifies that the specified value is a valid IP address.
-func IP(opCtx operation.Context, fldPath *field.Path, value, _ *string) field.ErrorList {
+// IPSloppy verifies that the specified value is a valid IP address, but allows
+// leading zeros on each octet value.  This should not be used for new APIs.
+func IPSloppy(opCtx operation.Context, fldPath *field.Path, value, _ *string) field.ErrorList {
```

Comment:
I might miss some context. May I know why this should not be used for new APIs?

Reply:
Go (and therefore k8s) used to allow strings like "001.002.003.004" to parse into IP addresses.  Kubernetes generally doesn't mutate user input, so we stored that in etcd.  The problem is that tools OUTSIDE of Go handle those in different ways.  Some reject them.  Some treat them as decimal.  Some treat them as octal.

The example I gave is not a big deal, but consider "010.020.030.040" -- is that "10.20.30.40" or is it "8.16.24.32" ?  Different tools disagree.

So Go broke compat and disallowed those values.  But we have them saved on disk!  If we just use Go's validation, we we fail to update objects which we previously allowed.  That's pretty high up on the list of things we don't want to do.

So we forked Go's old parser and dubbed it "sloppy".  We allow sloppy input in existing APIs but for new APis, by definition we do not have saved data, so we should not be sloppy.

Reply:
Good to know, thanks!